### DISCUSSION TITLE
After upgrading to Selenide version 7.0.4 from 6.18.0 my existing framework becomes unstable

### DISCUSSION THREAD

**Discussion by PrateekNehraFL** [selenide/selenide#2663] [Q&A] [answered]
After upgrading to Selenide version 7.0.4 from 6.18.0 my existing framework becomes unstable
I can see now List of SelenideElements needs to be changed to ElementsCollection 
![image](https://github.com/selenide/selenide/assets/82874048/b2d5eeba-3671-483f-b852-3ab3305bcb05)

So here is the piece of code in my framework that required that change
 
![image](https://github.com/selenide/selenide/assets/82874048/ed4a5060-dd85-4926-a006-f1d471018cc6)

After correcting it as per version 7.0 onwards the error is gone but now Elements collection doesn’t support stream() method
 
![image](https://github.com/selenide/selenide/assets/82874048/4fbeef80-d8f6-4b84-885a-be7ba1fc4040)
![image](https://github.com/selenide/selenide/assets/82874048/a815d6e3-66df-4e6e-b1ab-98b7d9ce8e36)

So to fix this up I tried this  
![image](https://github.com/selenide/selenide/assets/82874048/c2ac3c70-e33f-4f2c-a14d-ac811a723d2a)

Here we can see another error pops up due to my getText method  
![image](https://github.com/selenide/selenide/assets/82874048/deef62b9-cae2-46cf-a6f1-c06c71aa8bdb)

Here is my getText() method
![image](https://github.com/selenide/selenide/assets/82874048/01d8c710-59ed-4c2f-b14d-b38e8d436548)

 
Can someone please help me to fix this problem after version update?

Here is the POM.xml 
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>com.exadel</groupId>
    <artifactId>frontlineed-bdd</artifactId>
    <version>1.0-SNAPSHOT</version>

    <properties>
        <maven.compiler.source>17</maven.compiler.source>
        <maven.compiler.target>17</maven.compiler.target>
        <java.version>17</java.version>
        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <maven.surefire.version>3.0.0-M5</maven.surefire.version>
        <aspectj.version>1.9.6</aspectj.version>
        <allure-maven.version>2.11.2</allure-maven.version>
        <allure.version>2.17.3</allure.version>
        <cucumber.version>7.2.3</cucumber.version>
        <maven.compiler>3.8.0</maven.compiler>
        <junit.jupiter.version>5.8.2</junit.jupiter.version>
        <junit.platform.version>1.8.2</junit.platform.version>
        <logback.version>1.2.11</logback.version>
        <slf4j.version>1.7.36</slf4j.version>
        <log4j.version>2.17.2</log4j.version>
        <selenide.version>7.0.4</selenide.version>
        <selenium.version>4.16.1</selenium.version>
        <lombok.properties>1.18.24</lombok.properties>
        <rest.assured.version>5.1.1</rest.assured.version>
        <assertj.version>3.23.1</assertj.version>
        <ashot.version>1.5.4</ashot.version>
        <sikulix.version>2.0.5</sikulix.version>
        <reflections.version>0.9.12</reflections.version>
        <reportportal.version>5.0.1</reportportal.version>
        <allure.cucumber>2.17.3</allure.cucumber>
        <spring.framework>5.3.22</spring.framework>
        <pdfbox.version>2.0.26</pdfbox.version>
        <opencsv.version>5.7.0</opencsv.version>
        <nettycodec.version>4.1.103.Final</nettycodec.version>
    </properties>

    <dependencies>
        
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-api</artifactId>
            <version>${junit.jupiter.version}</version>
            <scope>test</scope>
        </dependency>

        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-engine</artifactId>
            <version>${junit.jupiter.version}</version>
            <scope>test</scope>
        </dependency>

        <dependency>
            <groupId>org.junit.platform</groupId>
            <artifactId>junit-platform-launcher</artifactId>
            <version>${junit.platform.version}</version>
            <scope>test</scope>
        </dependency>

        <dependency>
            <groupId>org.junit.platform</groupId>
            <artifactId>junit-platform-suite</artifactId>
            <version>${junit.platform.version}</version>
            <scope>test</scope>
        </dependency>

        <dependency>
            <groupId>ch.qos.logback</groupId>
            <artifactId>logback-core</artifactId>
            <version>${logback.version}</version>
        </dependency>
        <dependency>
            <groupId>ch.qos.logback</groupId>
            <artifactId>logback-classic</artifactId>
            <version>${logback.version}</version>
        </dependency>
        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
            <version>${slf4j.version}</version>
        </dependency>
        <dependency>
            <groupId>io.qameta.allure</groupId>
            <artifactId>allure-cucumber7-jvm</artifactId>
            <version>${allure.cucumber}</version>
        </dependency>
        <dependency>
            <groupId>com.codeborne</groupId>
            <artifactId>selenide</artifactId>
            <version>${selenide.version}</version>
        </dependency>
        <dependency>
            <groupId>io.cucumber</groupId>
            <artifactId>cucumber-java</artifactId>
            <version>${cucumber.version}</version>
        </dependency>
        <dependency>
            <groupId>io.cucumber</groupId>
            <artifactId>cucumber-core</artifactId>
            <version>${cucumber.version}</version>
        </dependency>
        <dependency>
            <groupId>io.cucumber</groupId>
            <artifactId>cucumber-junit</artifactId>
            <version>${cucumber.version}</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>io.cucumber</groupId>
            <artifactId>cucumber-junit-platform-engine</artifactId>
            <version>${cucumber.version}</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.seleniumhq.selenium</groupId>
            <artifactId>selenium-java</artifactId>
            <version>${selenium.version}</version>
        </dependency>
        <dependency>
            <groupId>io.rest-assured</groupId>
            <artifactId>rest-assured</artifactId>
            <version>${rest.assured.version}</version>
            <scope>compile</scope>
        </dependency>
        <dependency>
            <groupId>io.rest-assured</groupId>
            <artifactId>json-schema-validator</artifactId>
            <version>${rest.assured.version}</version>
        </dependency>
        <dependency>
            <groupId>org.assertj</groupId>
            <artifactId>assertj-core</artifactId>
            <version>${assertj.version}</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>ru.yandex.qatools.ashot</groupId>
            <artifactId>ashot</artifactId>
            <version>${ashot.version}</version>
        </dependency>
        <dependency>
            <groupId>com.sikulix</groupId>
            <artifactId>sikulixapi</artifactId>
            <version>${sikulix.version}</version>
        </dependency>
        <dependency>
            <groupId>org.reflections</groupId>
            <artifactId>reflections</artifactId>
            <version>${reflections.version}</version>
        </dependency>
        <dependency>
            <groupId>com.epam.reportportal</groupId>
            <artifactId>agent-java-cucumber5</artifactId>
            <version>${reportportal.version}</version>
        </dependency>
        <dependency>
            <groupId>org.projectlombok</groupId>
            <artifactId>lombok</artifactId>
            <version>${lombok.properties}</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>org.apache.logging.log4j</groupId>
            <artifactId>log4j-api</artifactId>
            <version>${log4j.version}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.logging.log4j</groupId>
            <artifactId>log4j-core</artifactId>
            <version>${log4j.version}</version>
        </dependency>
        <dependency>
            <groupId>io.qameta.allure</groupId>
            <artifactId>allure-selenide</artifactId>
            <version>${allure.version}</version>
            <exclusions>
                <exclusion>
                    <artifactId>selenide</artifactId>
                    <groupId>com.codeborne</groupId>
                </exclusion>
                <exclusion>
                    <artifactId>jackson-annotations</artifactId>
                    <groupId>com.fasterxml.jackson.core</groupId>
                </exclusion>
            </exclusions>
        </dependency>
        <dependency>
            <groupId>org.springframework</groupId>
            <artifactId>spring-core</artifactId>
            <version>${spring.framework}</version>
        </dependency>
        <dependency>
            <groupId>org.springframework</groupId>
            <artifactId>spring-context</artifactId>
            <version>${spring.framework}</version>
        </dependency>
        <dependency>
            <groupId>org.springframework</groupId>
            <artifactId>spring-jdbc</artifactId>
            <version>${spring.framework}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.pdfbox</groupId>
            <artifactId>pdfbox</artifactId>
            <version>${pdfbox.version}</version>
        </dependency>
        <dependency>
            <groupId>com.opencsv</groupId>
            <artifactId>opencsv</artifactId>
            <version>${opencsv.version}</version>
        </dependency>
        
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-codec</artifactId>
            <version>${nettycodec.version}</version>
        </dependency>

    </dependencies>

    <profiles>
        <profile>
            <id>tests</id>
            <activation>
                <activeByDefault>true</activeByDefault>
            </activation>
            <properties>
                <runner>**/RunCukesTest.java</runner>
            </properties>
        </profile>
    </profiles>
    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>3.0.0-M7</version>
                <configuration>
                    <parallel>methods</parallel>

                    <threadCount>20</threadCount>
                    <properties>
                        <configurationParameters>
                            cucumber.junit-platform.naming-strategy=long
                        </configurationParameters>
                    </properties>
                    <includes>
                        <include>**/*Run*.java</include>
                    </includes>
                    <testFailureIgnore>true</testFailureIgnore>
                    <argLine>
                        -javaagent:"${settings.localRepository}/org/aspectj/aspectjweaver/${aspectj.version}/aspectjweaver-${aspectj.version}.jar"
                    </argLine>
                </configuration>
                <dependencies>
                    <dependency>
                        <groupId>org.aspectj</groupId>
                        <artifactId>aspectjweaver</artifactId>
                        <version>${aspectj.version}</version>
                    </dependency>
                </dependencies>
            </plugin>
        </plugins>
    </build>
</project>

**Comment by rosolko** [comment, marked as answer]
As for collections you need to call `asFixedIterable/asDynamicIterable` method on collection before `stream` call.
As for `getText()` - it's not working because you call `texts()` method, which execute `getText()` on each element and returns a stream of string for you.

**Reply by PrateekNehraFL** [reply]
Thanks a lot for quick response it did worked @rosolko . However, another minor thing comes up for which I am quite confused to figure out the solution . Here now how to cast ElementsCollection to ArrayList ? 
![image](https://github.com/selenide/selenide/assets/82874048/78182a09-adbf-4658-a66c-120ebc375756)

**Reply by PrateekNehraFL** [reply]
Got it it worked this way, thanks for all your help !
![image](https://github.com/selenide/selenide/assets/82874048/b266b478-b525-4758-935f-19264fcc6f80)

**Comment by asolntsev** [comment]
@PrateekNehraFL I recommend you to throw away method `getAnyElementsListTexts()`. It returns list of texts, but do you do with these texts? Pretty sure, you want to verify those texts, right? 

Then just verify:
```java
  getListElement(listName).shouldHave(texts("one", "two", "three"));
```

All these iterators, streams, dynamic iterables are... evil.

**Reply by PrateekNehraFL** [reply]
@asolntsev thanks for the information !
One more problem that I am facing every now and then after updating to Selenium version 4.18.1 and Selenide version 7.0.4 
Timed out receiving message from renderer

![image](https://github.com/selenide/selenide/assets/82874048/c757aa43-215e-44ae-8e3d-d737785a7438)

**Reply by BorisOsipov** [reply]
@PrateekNehraFL You can try to increase `com.codeborne.selenide.Configuration#pageLoadStrategy`. In the old version, the default was larger than in v7 AFAIK

**Reply by PrateekNehraFL** [reply]
@BorisOsipov @asolntsev many a times the test fails with timeout receiving message from renderer with less than a second despite pageLoadTimeout set to 60 seconds
![image](https://github.com/selenide/selenide/assets/82874048/2783e911-13d0-44f8-97fa-d604c3cf320a)

**Reply by PrateekNehraFL** [reply]
@asolntsev any idea how to fix this timed out renderer ?

**Reply by asolntsev** [reply]
@PrateekNehraFL It seems this question is not related to Selenide. It's rather about your environment (browsers, dockers, network etc.)
